test/select.jl

```diff
@@ -0,0 +1,838 @@
+        cdf = copy(df)
+        select!(cdf, :x => ByRow(x -> retval))
+        @test cdf == DataFrame(x_function = [retval])
+    end
+
+    for retval in [(1, 2), ones(2,2,2)]
+        @test select(df, :x => x -> retval) == DataFrame(x_function = [retval])
+        @test select(df, :x => ByRow(x -> retval)) == DataFrame(x_function = [retval])
+        cdf = copy(df)
+        select!(cdf, :x => x -> retval)
+        @test cdf == DataFrame(x_function = [retval])
+        cdf = copy(df)
+        select!(cdf, :x => ByRow(x -> retval))
+        @test cdf == DataFrame(x_function = [retval])
+    end
+end
+
+@testset "select and select! empty selection" begin
+    df = DataFrame(rand(10, 4))
+    x = [1,2,3]
+
+    @test select(df, r"z") == DataFrame()
+    @test select(df, r"z" => () -> x) == DataFrame(_function=x)
+    @test select(df, r"z" => () -> x)[!, 1] === x # no copy even for copycols=true
+    @test_throws MethodError select(df, r"z" => x -> 1)
+    @test_throws ArgumentError select(df, r"z" => ByRow(rand))
+
+    @test select(df, r"z", copycols=false) == DataFrame()
+    @test select(df, r"z" => () -> x, copycols=false) == DataFrame(_function=x)
+    @test select(df, r"z" => () -> x, copycols=false)[!, 1] === x
+    @test_throws MethodError select(df, r"z" => x -> 1, copycols=false)
+    @test_throws ArgumentError select(df, r"z" => ByRow(rand), copycols=false)
+
+    @test_throws MethodError select!(df, r"z" => x -> 1)
+    @test_throws ArgumentError select!(df, r"z" => ByRow(rand))
+    @test_throws ErrorException select!(df, r"z" => () -> x, copycols=false)
+    select!(df, r"z" => () -> x)
+    @test df == DataFrame(_function=x)
+end
+
+@testset "wrong selection patterns" begin
+    df = DataFrame(rand(10, 4))
+
+    @test_throws ArgumentError select(df, "z")
+    @test_throws ArgumentError select(df, "z" => :x1)
+    @test_throws ArgumentError select(df, "z" => identity)
+    @test_throws ArgumentError select(df, "z" => identity => :x1)
+end
+
+@testset "select and select! duplicates" begin
+    df = DataFrame(rand(10, 4))
+    df_ref = copy(df)
+
+    @test_throws ArgumentError select(df, :x1, :x1)
+    @test_throws ArgumentError select(df, :x1, :x5)
+    @test select(df, :x2, r"x", :x1, :) == df[:, [:x2, :x1, :x3, :x4]]
+
+    @test_throws ArgumentError select(df, :x1, :x2 => :x1)
+    @test_throws ArgumentError select(df, :x3 => :x1, :x2 => :x1)
+    @test_throws ArgumentError select(df, :x1, :x2 => identity => :x1)
+    @test_throws ArgumentError select(df, :x1 => :x1, :x2 => identity => :x1)
+    @test_throws ArgumentError select(df, :x3 => identity => :x1, :x2 => identity => :x1)
+    @test select(df, [:x1], :x2 => :x1) == DataFrame(x1 = df.x2)
+
+    @test_throws ArgumentError select!(df, :x1, :x1)
+    @test_throws ArgumentError select!(df, :x1, :x5)
+    @test df == df_ref
+
+    select!(df, :x2, r"x", :x1, :)
+    @test df == df_ref[:, [:x2, :x1, :x3, :x4]]
+end
+
+@testset "SubDataFrame selection" begin
```

Comment:
Same remark about loop.

Reply:
The tests are not 100% identical here, and I prefer to keep tests grouped by data frame type, as then it was easier for me to make sure I cover all cases.